crates/kitsune_p2p/types/src/tx2/tx2_promote.rs

```diff
@@ -0,0 +1,701 @@
+    writer_bucket: ResourceBucket<WriteChan>,
+}
+
+struct PromoteConHnd(Arc<Share<PromoteConInner>>, Uniq);
+
+impl std::fmt::Debug for PromoteConHnd {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let a = self
+            .0
+            .share_mut(|i, _| i.con.remote_addr().map(|a| a.to_string()))
+            .unwrap_or_else(|_| "[closed]".to_string());
+        f.debug_tuple("ConHnd").field(&a).finish()
+    }
+}
+
+async fn in_chan_recv_logic(
+    inner: Arc<Share<PromoteConInner>>,
+    raw_con: Arc<dyn ConAdapt>,
+    con: ConHnd,
+    writer_bucket: ResourceBucket<WriteChan>,
+    logic_hnd: LogicChanHandle<EpEvent>,
+    in_chan_recv: Box<dyn InChanRecvAdapt>,
+) {
+    let url = match con.remote_addr() {
+        Err(e) => {
+            // unable to determine remote url for connection
+
+            let reason = format!("{:?}", e);
+
+            // TODO - standardize codes?
+            close_inner(&inner, 500, &reason).await;
+
+            // don't even start the loops
+            return;
+        }
+        Ok(url) => url,
+    };
+
+    let url = &url;
+    let con = &con;
+    let logic_hnd = &logic_hnd;
+    let inner = &inner;
+
+    let recv_fut = in_chan_recv
+        .for_each_concurrent(CHANNELS_PER_CONNECTION, move |chan| async move {
+            let mut chan = match chan.await {
+                Err(e) => {
+                    // unable to resolve incoming channel
+                    // shut down the connection
+
+                    let reason = format!("{:?}", e);
+
+                    // TODO - standardize codes?
+                    close_inner(inner, 500, &reason).await;
+
+                    // exit the loop
+                    return;
+                }
+                Ok(c) => c,
+            };
+            loop {
+                let r = chan
+                    .read(KitsuneTimeout::from_millis(MAX_READ_TIMEOUT))
+                    .await;
+                let (msg_id, data) = match r {
+                    Err(e) if *e.kind() == KitsuneErrorKind::Closed => {
+                        // this channel was closed - exit the loop
+                        // allowing another channel to be processed.
+                        return;
+                    }
+                    Err(e) => {
+                        // unrecoverable error - shut down the connection
+
+                        let reason = format!("{:?}", e);
+
+                        // TODO - standardize codes?
+                        close_inner(inner, 500, &reason).await;
+
+                        // exit the loop
+                        return;
+                    }
+                    Ok(r) => r,
+                };
+
+                let data = EpIncomingData {
+                    url: url.clone(),
+                    con: con.clone(),
+                    msg_id,
+                    data,
+                };
+                if logic_hnd.emit(EpEvent::IncomingData(data)).await.is_err() {
```

Comment:
Any reason to make this error silent? Is it worth logging?

Reply:
added:

```rust
                    // the only reason this will error is if our
                    // endpoint is shut down, in which case we
                    // no longer care about the error.
```